### PR TITLE
 Add new feature for supporting WebAssembly artifact

### PR DESCRIPTION
**Pull request by ln23415**
This PR enable harbor to identify WebAssembly artifact. User can use wasm-to-oci tool pushing WebAssembly artifact to harbor, or push an "module.wasm.image/variant=compat" annotated WebAssembly artifact to harbor.

### CODE REVIEW

**Review by Vad1mo — COMMENTED** [review]
Thank you for your contribution, this looks neat. 
Would you be able to provide some tests?

Did you have a look at https://github.com/goharbor/harbor/discussions/16775#discussioncomment-2839939

**Comment by wy65701436** [comment]
@ln23415 could you please help to make the CI pass?

**Comment by wy65701436** [comment]
can we have the offical icon? Instead of the screenshot.

File: src/controller/artifact/processor/wasm/wasm.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
+// Copyright Project Harbor Authors
```
Comment (wy65701436) [inline src/controller/artifact/processor/wasm/wasm.go:1]: please add UT for wasm

File: src/controller/artifact/controller.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ import (
 	"context"
 	stderrors "errors"
 	"fmt"
+	"github.com/goharbor/harbor/src/controller/artifact/processor/wasm"
```
Comment (wy65701436) [inline src/controller/artifact/controller.go:22]: please revise the import, it has to move this line into github.com/goharbor/harbor group

File: src/controller/artifact/processor/wasm/wasm.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		author := config.Author
+		if len(author) == 0 && len(config.Config.Labels) > 0 {
+			author = config.Config.Labels["maintainer"]
+		}
+		art.ExtraAttrs["author"] = author
+	}else {
+
+		// for wasm-to-oci way
+		art.ExtraAttrs["manifest.config.mediaType"] = mediaType
+		art.ExtraAttrs["manifest.layers.mediaType"] = manifest.Layers[0].MediaType
```
Comment (wy65701436) [inline src/controller/artifact/processor/wasm/wasm.go:89]: please add array check before using.

File: src/controller/artifact/abstractor.go [hunk 4]
```diff
@@ -126,6 +126,11 @@ func (a *abstractor) abstractManifestV2Metadata(artifact *artifact.Artifact, con
 	}
 	// use the "manifest.config.mediatype" as the media type of the artifact
 	artifact.MediaType = manifest.Config.MediaType
+
+	if manifest.Annotations["module.wasm.image/variant"]=="compat" || manifest.Annotations["run.oci.handler"]=="wasm" {
```
Comment (stonezdj) [inline src/controller/artifact/abstractor.go:130]: extract "module.wasm.image/variant" and "run.oci.handler" as constant?

File: src/controller/artifact/processor/wasm/wasm.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+	art.ExtraAttrs = map[string]interface{}{}
+	art.ExtraAttrs["ArtifactType"] = "WebAssembly"
+
+	manifest := &v1.Manifest{}
+	if err := json.Unmarshal(manifest_body, manifest); err != nil {
+		return err
+	}
+
+	if manifest.Annotations["module.wasm.image/variant"]=="compat" || manifest.Annotations["run.oci.handler"]=="wasm" {
```
Comment (stonezdj) [inline src/controller/artifact/processor/wasm/wasm.go:63]: extract constant

File: src/controller/artifact/processor/wasm/wasm.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+// manifestV2Processor processes image with OCI manifest and docker v2 manifest
+type manifestV2Processor struct {
+	*base.ManifestProcessor
+}
+
+func (m *manifestV2Processor) AbstractMetadata(ctx context.Context, art *artifact.Artifact, manifest_body []byte) error {
+
+	art.ExtraAttrs = map[string]interface{}{}
+	art.ExtraAttrs["ArtifactType"] = "WebAssembly"
```
Comment (ywk253100) [inline src/controller/artifact/processor/wasm/wasm.go:56]: Why put the `ArtifactType` into the `ExtraAttrs`? The type is already a standard field for artifact

File: src/controller/artifact/processor/wasm/wasm.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	if manifest.Annotations["module.wasm.image/variant"]=="compat" || manifest.Annotations["run.oci.handler"]=="wasm" {
+
+		// for annotation way
+		config := &v1.Image{}
+		if err := m.UnmarshalConfig(ctx, art.RepositoryName, manifest_body, config); err != nil {
+			return err
+		}
+		art.ExtraAttrs["manifest.config.mediaType"] = manifest.Config.MediaType
+		if art.ExtraAttrs == nil {
+			art.ExtraAttrs = map[string]interface{}{}
```
Comment (ywk253100) [inline src/controller/artifact/processor/wasm/wasm.go:72]: You have already put values into the `ExtraAttrs` before checking whether it's nil or not.

File: src/controller/artifact/processor/wasm/wasm.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		art.ExtraAttrs["created"] = config.Created
```
Comment (ywk253100) [inline src/controller/artifact/processor/wasm/wasm.go:74]: Are those `created`, `architecture`, `os`, `config`, etc the attributes of WebAssembly? If they are not, I don't think the `AbstractMetadata()` is necessary

File: src/controller/artifact/processor/wasm/wasm.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+		art.ExtraAttrs["manifest.config.mediaType"] = mediaType
+		art.ExtraAttrs["manifest.layers.mediaType"] = manifest.Layers[0].MediaType
+		art.ExtraAttrs["org.opencontainers.image.title"] = manifest.Layers[0].Annotations["org.opencontainers.image.title"]
+	}
+	return nil
+}
+
+func (m *manifestV2Processor) AbstractAddition(ctx context.Context, artifact *artifact.Artifact, addition string) (*processor.Addition, error) {
+
+	if addition != AdditionTypeBuildHistory {
```
Comment (ywk253100) [inline src/controller/artifact/processor/wasm/wasm.go:97]: Similar question, the WebAssembly also has the build history concept, right?

File: src/controller/artifact/abstractor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ln23415) [reply src/controller/artifact/abstractor.go:130]: yes, in this annotation wasm fashion, the media-type of artifact is "application/vnd.docker.distribution.manifest.v2+json", which conflicts with docker image, and that is why I alter it, handing the artifact to wasm processor

File: src/controller/artifact/abstractor.go [hunk 10]
```diff
@@ -126,6 +126,11 @@ func (a *abstractor) abstractManifestV2Metadata(artifact *artifact.Artifact, con
...
+		artifact.MediaType = "application/vnd.wasm.config.v1+json"
```
[the remark this answers is not in the document]
Comment (ln23415) [reply src/controller/artifact/abstractor.go:131]: new PR will fix this

File: src/controller/artifact/controller.go (hunk 2 above)
Comment (ln23415) [reply src/controller/artifact/controller.go:22]: new PR will fix this

File: src/controller/artifact/processor/wasm/wasm.go (hunk 1 above)
Comment (ln23415) [reply src/controller/artifact/processor/wasm/wasm.go:1]: new PR will fix this

File: src/controller/artifact/processor/wasm/wasm.go (hunk 6 above)
Comment (ln23415) [reply src/controller/artifact/processor/wasm/wasm.go:56]: new PR will fix this

File: src/controller/artifact/processor/wasm/wasm.go (hunk 5 above)
Comment (ln23415) [reply src/controller/artifact/processor/wasm/wasm.go:63]: new PR will fix this

File: src/controller/artifact/processor/wasm/wasm.go (hunk 7 above)
Comment (ln23415) [reply src/controller/artifact/processor/wasm/wasm.go:72]: new PR will fix this

File: src/controller/artifact/processor/wasm/wasm.go (hunk 8 above)
Comment (ln23415) [reply src/controller/artifact/processor/wasm/wasm.go:74]: yes, these flags are the attributes of WebAssembly l in annotation fashion wasm artifact.

File: src/controller/artifact/processor/wasm/wasm.go (hunk 3 above)
Comment (ln23415) [reply src/controller/artifact/processor/wasm/wasm.go:89]: new PR will fix this

File: src/controller/artifact/processor/wasm/wasm.go (hunk 9 above)
Comment (ln23415) [reply src/controller/artifact/processor/wasm/wasm.go:97]: yes, it happens in annotation fashion wasm artifact

**Comment by wy65701436** [comment]
@ln23415 Please address these comments in this PR instead of starting a new thread, thanks.

**Comment by wy65701436** [comment]
> > 
> 
> Do you mean the WebAssembly icon? this is the official icon.

yes, I mean the official icon, would you mind share the uri of this icon?

**Review by wy65701436 — APPROVED** [review]
lgtm

**Comment by Vad1mo** [comment]
While I like the general idea, however I would favor a more robust solution supporting differents image types. 

See https://github.com/goharbor/harbor/discussions/16775#discussioncomment-2839939

**Pull request merged**